### ISSUE TITLE
[$250] [Performance] Investigate `react-freeze` to prevent updates to the LHN / Report page - reported by @vladamx

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#11589]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

inefficient React component rendering

Specifically, updates to the SidebarLinks and ReportScreen can cause performance to drop when transitioning between these screens on Androids. Less rendering in theory should yield less work and faster transitions without any changes in behavior for the user.

## What is the impact of this on end-users?

Janky transition, slow interactions when chat switching

## List any benchmarks that show the severity of the issue

cc @vladamx

## Proposed solution (if any)

It sounds like the proposed solution is to try to upgrade to a more recent version of `react-native-screens` and enable the screen freezing options in the correct places.

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)

cc @vladamx

## Platform:
Where is this issue occurring?

- [x] Web
- [x] iOS
- [x] Android
- [x] Desktop App
- [x] Mobile Web

**Version Number:** 1.2.11-2
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by marcaaron** [comment]
@Luke9389 @mananjadhav I'm suggesting we give first dibs to @vladamx to fix this issue. They are working on another small issue atm, but I think totally reasonable for them to work on this one as well.

**Comment by marcaaron** [comment]
@vladamx maybe you can consider adding some benchmarks here with your proposal. I am guessing that chat switching should improve a bit on Android and would be pretty obvious after the improvement. You can filter for this metric in the JS console like this:

![Screen Shot 2022-10-04 at 7 58 00 AM](https://user-images.githubusercontent.com/32969087/193892000-cfcaa510-391c-40bd-a99f-0b4f80bd819c.png)

**Comment by mallenexpensify** [comment]
@marcaaron few questions: 

1. Is there a Slack link to a reported bug? 
2. Should we remove the `Help Wanted` label?
3. Should we wait for @vladamx to express interest or submit a proposal before hiring? 

Thx

**Comment by marcaaron** [comment]
1. https://expensify.slack.com/archives/C01GTK53T8Q/p1664892711333199
2. No - let's wait for @vladamx
3. Yes

**Comment by Luke9389** [comment]
OK cool! I'm down for this plan.

**Comment by vladamx** [comment]
Proposal

In order to improve general responsiveness of the app and specifically when switching reports the idea is to only re-render screens when they are visible to user. This will avoid needless work caused by shared state.

Implementation details:

Upgrade to a more recent version of react-native-screens and enable the screen freezing options in the correct places by using React Navigation API https://reactnavigation.org/docs/drawer-navigator/#freezeonblur

**Comment by mananjadhav** [comment]
I went through the conversation on the slack and @vladamx's proposal here. @vladamx Is it possible for you to elaborate on the implementation details? Which part of the codebase? Which screens are we talking about to begin with?

and as @marcaaron mentioned earlier if we can get some benchmarks along with the proposal, it'll be of great help.

**Comment by vladamx** [comment]
@mananjadhav this is primarily for sidebar and report to not influence eachother.

There are a lot of metrics that will be influenced one of which i already sent.

I can also include metric for switching reports right away, if you find it valuable.

**Comment by vladamx** [comment]
Ok. I will provide the metric and code snippet

**Comment by vladamx** [comment]
Quick update on this one,

I upgraded drawer, stack, react native screens and enabled freeze globally with enableFreeze(true).

However i am still getting re-renders in the profiler that i would not expect to happen. Even if i go deep in settings stack it still re-renders report.

I will need to dig deeper and try to use react-freeze API directly which is just a component you can wrap your content with and suspend rendering by passing a boolean.

Also, after working a bit with this lib it might not work well with drawer content(which was my initial proposal for optimization) since its not a screen per say.

**Comment by marcaaron** [comment]
I have been experimenting as well. What I am seeing is that if we use `<Freeze />` on some elements then they will appear with a white space when `freeze={true}`. That doesn't work so great with the drawer navigator since the expected behavior is that if you are swiping back or forward on the drawer that you'd be able to see some preview of the content while transitioning.

That said, it might be more important to have the chat switch happen quickly than to have the preview happen. More context in this thread:

https://expensify.slack.com/archives/C01GTK53T8Q/p1665169423327639

**Comment by vladamx** [comment]
**Proposal**

Ok, after some internal discussions and alignments with @marcaaron I came up with PR Proposal https://github.com/Expensify/App/pull/11798

Basically I use react-freeze to suspend rendering in SidebarLinks and ReportScreen when they are not active by setting the freeze prop to depend on the drawer state. I do this only on mobile since on desktop they are both active at the same time.

https://user-images.githubusercontent.com/11381171/195720261-5b58013a-f3fd-4c55-ac68-5cd76308e294.mov

By going with this approach switch between reports is automatic - There is no lag in transition. It takes some time for chat data to show since we are suspending the screen. 

Plus, shared state updates are now not triggering deep re-renderings when ReportScreen or SidebarLinks are not visible/active.

cc @mananjadhav

**Comment by marcaaron** [comment]
We are getting pretty close to merging [a PR that ended up implementing `react-freeze`](https://github.com/Expensify/App/pull/11550). I think what we should do is:

- Close this issue
- Pay out @vladamx for this idea + the bonus for reporting the issue. 

Ultimately, it was their idea to use `react-freeze` and I just ended up needing it for another change I am working on. Creating another PR at this point is maybe not worth everyone's time, but the contribution here was valuable. Alternatively, we could approve the above proposal and merge in the changes - but I'd prefer not to block my other PR on that process.

**Comment by laurenreidexpensify** [comment]
@kevinksullivan sorry I've been OOO for a family emergency and am only back on 20 Oct, so can't action, Can you please create a job on upwork and pay @vladamx before closing this? Thanks!

**Comment by Luke9389** [comment]
Still waiting on the upwork job to be created.

**Comment by mallenexpensify** [comment]
Job created!! 
https://www.upwork.com/jobs/~012d0f5ce31935509e

**Comment by kevinksullivan** [comment]
Offer is out @vladamx ! Let me know when you'e accepted.

**Comment by vladamx** [comment]
@kevinksullivan Is reporting bonus separate payment on top of job offer(250)?
